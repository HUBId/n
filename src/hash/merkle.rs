//! BLAKE3-based binary Merkle tree implementation used by the proving system.
//!
//! The tree adheres to the specification circulated with the repository:
//!
//! * Inner nodes hash the concatenation of their two children in order.
//! * Leaves are hashed as `BLAKE3(u32_le(len) || payload)`.
//! * Missing children on the right-hand side are padded with a fixed
//!   `EMPTY` digest derived from the string `"RPP-MERKLE-EMPTY\0"`.
//! * Authentication paths serialise an index byte followed by the sibling
//!   digest ordered by its position (0 or 1) within the parent.
//!
//! The module provides helpers to build Merkle trees, derive authentication
//! paths and recompute the root from a path.  Errors are surfaced when a
//! caller attempts to verify malformed paths (e.g. mismatched padding or
//! tampered length prefixes).

use crate::hash::{hash, Hasher};
use core::fmt;

/// Number of children per internal node.
const ARITY: usize = 2;

/// Size of a digest emitted by the tree (BLAKE3 output size).
pub const DIGEST_SIZE: usize = 32;

/// Canonical digest for an empty child.
pub const EMPTY_DIGEST: [u8; DIGEST_SIZE] = [
    0x10, 0x68, 0x98, 0xac, 0xc4, 0xcc, 0xf4, 0xd1, 0x1d, 0x07, 0x4a, 0x00, 0x09, 0xbc, 0x4a, 0x8e,
    0x56, 0x29, 0x03, 0x5a, 0xf3, 0x05, 0x46, 0xdc, 0xdb, 0xff, 0xba, 0x19, 0x77, 0xa7, 0x59, 0x61,
];

/// Digest binding the documented Merkle layout for parameter commitments.
pub const MERKLE_SCHEME_ID: [u8; DIGEST_SIZE] = [
    0xa0, 0x02, 0xd1, 0xc1, 0x58, 0xcd, 0x9b, 0x91, 0x2d, 0x51, 0x38, 0x67, 0x50, 0x7b, 0xcd, 0x19,
    0x5a, 0xe8, 0x15, 0xfa, 0x16, 0xdb, 0x7b, 0x3a, 0xc2, 0x5f, 0xae, 0x6f, 0x6b, 0xd9, 0x13, 0xa4,
];

/// Index of a child within a binary node (stored as little-endian byte in proofs).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct MerkleIndex(pub u8);

impl MerkleIndex {
    /// Maximum allowed index for the binary fan-out.
    pub const MAX: u8 = 1;
}

/// Path element capturing siblings and the caller position.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MerklePathElement {
    /// Position of the caller node within the parent (`0..=1`).
    pub index: MerkleIndex,
    /// Sibling hash ordered from left (0) to right (1).
    pub siblings: [[u8; DIGEST_SIZE]; ARITY - 1],
}

/// Errors reported while verifying or constructing Merkle proofs.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MerkleError {
    /// The u32 little-endian length prefix disagrees with the payload size.
    ErrMerkleLeafLength,
    /// Missing leaves on the right-hand side were not padded with `EMPTY`.
    ErrMerkleEmptyPadding,
    /// Sibling order within a Merkle proof was inconsistent with the specification.
    ErrMerkleSiblingOrder,
    /// Encountered an invalid index byte or inconsistent path structure.
    ErrPathIndexByte,
}

impl fmt::Display for MerkleError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MerkleError::ErrMerkleLeafLength => write!(f, "ErrFRIPathInvalid: leaf length"),
            MerkleError::ErrMerkleEmptyPadding => write!(f, "ErrFRIPathInvalid: right padding"),
            MerkleError::ErrMerkleSiblingOrder => write!(f, "ErrFRIPathInvalid: sibling order"),
            MerkleError::ErrPathIndexByte => write!(f, "ErrFRIPathInvalid: index byte"),
        }
    }
}

impl std::error::Error for MerkleError {}

/// Convenience wrapper for a binary BLAKE3 Merkle tree.
#[derive(Debug, Clone)]
pub struct Blake3MerkleTree {
    levels: Vec<Vec<[u8; DIGEST_SIZE]>>,
    leaf_count: usize,
}

impl Blake3MerkleTree {
    /// Builds a Merkle tree from canonical leaf encodings.
    ///
    /// Each leaf must follow the framing `len: u32 (little-endian) || payload`.
    pub fn from_leaves<I, B>(leaves: I) -> Result<Self, MerkleError>
    where
        I: IntoIterator<Item = B>,
        B: AsRef<[u8]>,
    {
        let mut hashed = Vec::new();
        let mut leaf_count = 0usize;
        for leaf in leaves {
            let leaf_bytes = leaf.as_ref();
            hashed.push(hash_leaf(leaf_bytes)?);
            leaf_count += 1;
        }

        if leaf_count == 0 {
            return Ok(Self {
                levels: vec![vec![EMPTY_DIGEST]],
                leaf_count: 0,
            });
        }

        let mut levels = Vec::new();
        levels.push(hashed.clone());
        let mut current = hashed;

        while current.len() > 1 {
            let mut next = Vec::with_capacity((current.len() + ARITY - 1) / ARITY);
            for chunk in current.chunks(ARITY) {
                let mut children = [[0u8; DIGEST_SIZE]; ARITY];
                for (position, child) in children.iter_mut().enumerate() {
                    *child = if position < chunk.len() {
                        chunk[position]
                    } else {
                        EMPTY_DIGEST
                    };
                }
                next.push(hash_internal(&children));
            }
            levels.push(next.clone());
            current = next;
        }

        Ok(Self { levels, leaf_count })
    }

    /// Returns the Merkle root digest.
    pub fn root(&self) -> [u8; DIGEST_SIZE] {
        self.levels
            .last()
            .and_then(|level| level.first().copied())
            .unwrap_or(EMPTY_DIGEST)
    }

    /// Number of leaves provided when building the tree (after implicit padding).
    pub fn leaf_count(&self) -> usize {
        self.leaf_count
    }

    /// Generates an authentication path for the leaf at `index`.
    pub fn open(&self, index: usize) -> Result<Vec<MerklePathElement>, MerkleError> {
        if index >= self.leaf_count {
            return Err(MerkleError::ErrPathIndexByte);
        }

        let mut path = Vec::with_capacity(self.levels.len().saturating_sub(1));
        let mut current_index = index;

        for level in 0..self.levels.len().saturating_sub(1) {
            let nodes = &self.levels[level];
            let parent_index = current_index / ARITY;
            let position = current_index % ARITY;
            let chunk_base = parent_index * ARITY;
            let chunk_len = nodes.len().saturating_sub(chunk_base).min(ARITY);

            let sibling = if position == 0 {
                if chunk_len > 1 {
                    nodes[chunk_base + 1]
                } else {
                    EMPTY_DIGEST
                }
            } else if chunk_len > 1 {
                nodes[chunk_base]
            } else {
                EMPTY_DIGEST
            };

            path.push(MerklePathElement {
                index: MerkleIndex(position as u8),
                siblings: [sibling],
            });

            current_index /= ARITY;
        }

        Ok(path)
    }
}

/// Encodes raw payload bytes into the canonical leaf representation.
pub fn encode_leaf(payload: &[u8]) -> Vec<u8> {
    let mut encoded = Vec::with_capacity(4 + payload.len());
    encoded.extend_from_slice(&(payload.len() as u32).to_le_bytes());
    encoded.extend_from_slice(payload);
    encoded
}

/// Hashes a leaf using the canonical framing rules while checking the length prefix.
pub fn hash_leaf(encoded_leaf: &[u8]) -> Result<[u8; DIGEST_SIZE], MerkleError> {
    if encoded_leaf.len() < 4 {
        return Err(MerkleError::ErrMerkleLeafLength);
    }

    let mut len_bytes = [0u8; 4];
    len_bytes.copy_from_slice(&encoded_leaf[..4]);
    let declared_len = u32::from_le_bytes(len_bytes) as usize;
    let payload = &encoded_leaf[4..];

    if declared_len != payload.len() {
        return Err(MerkleError::ErrMerkleLeafLength);
    }

    Ok(hash(encoded_leaf).into())
}

/// Hashes two child digests into their parent digest.
pub fn hash_internal(children: &[[u8; DIGEST_SIZE]; ARITY]) -> [u8; DIGEST_SIZE] {
    let mut hasher = Hasher::new();
    for child in children {
        hasher.update(child);
    }
    hasher.finalize().into()
}

/// Recomputes the Merkle root from a leaf and its authentication path.
///
/// * `leaf`: canonical encoding `len || payload`.
/// * `index`: position of the leaf within the tree (0-based).
/// * `leaf_count`: total number of leaves committed by the tree.
/// * `path`: authentication path (leaf to root order).
pub fn compute_root_from_path(
    leaf: &[u8],
    index: usize,
    leaf_count: usize,
    path: &[MerklePathElement],
) -> Result<[u8; DIGEST_SIZE], MerkleError> {
    if leaf_count == 0 || index >= leaf_count {
        return Err(MerkleError::ErrPathIndexByte);
    }

    let mut hash = hash_leaf(leaf)?;
    let mut current_index = index;
    let mut nodes_in_level = leaf_count;

    for element in path {
        if element.index.0 > MerkleIndex::MAX {
            return Err(MerkleError::ErrPathIndexByte);
        }
        let expected_position = current_index % ARITY;
        if element.index.0 as usize != expected_position {
            return Err(MerkleError::ErrPathIndexByte);
        }

        let parent_index = current_index / ARITY;
        let chunk_base = parent_index * ARITY;
        let chunk_len = nodes_in_level.saturating_sub(chunk_base).min(ARITY);

        let mut children = [[0u8; DIGEST_SIZE]; ARITY];
        children[expected_position] = hash;
        let sibling = element
            .siblings
            .get(0)
            .copied()
            .ok_or(MerkleError::ErrPathIndexByte)?;
        let sibling_position = expected_position ^ 1;
        if sibling_position >= chunk_len && sibling != EMPTY_DIGEST {
            return Err(MerkleError::ErrMerkleEmptyPadding);
        }
        children[sibling_position] = if sibling_position < chunk_len {
            sibling
        } else {
            EMPTY_DIGEST
        };

        hash = hash_internal(&children);
        current_index = parent_index;
        nodes_in_level = (nodes_in_level + ARITY - 1) / ARITY;
    }

    if current_index != 0 {
        return Err(MerkleError::ErrPathIndexByte);
    }

    Ok(hash)
}

/// Verifies a Merkle path against an expected root digest.
pub fn verify_path(
    leaf: &[u8],
    index: usize,
    leaf_count: usize,
    path: &[MerklePathElement],
    expected_root: &[u8; DIGEST_SIZE],
) -> Result<(), MerkleError> {
    let computed = compute_root_from_path(leaf, index, leaf_count, path)?;
    if &computed != expected_root {
        return Err(MerkleError::ErrMerkleSiblingOrder);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_digest_matches_reference() {
        assert_eq!(EMPTY_DIGEST, hash(b"RPP-MERKLE-EMPTY\0").into_bytes());
    }

    #[test]
    fn empty_tree_root_is_empty() {
        let leaves: Vec<Vec<u8>> = Vec::new();
        let tree = Blake3MerkleTree::from_leaves(leaves).expect("tree");
        assert_eq!(tree.leaf_count(), 0);
        assert_eq!(tree.root(), EMPTY_DIGEST);
    }

    #[test]
    fn build_path_verify_roundtrip_ok() {
        let payloads = vec![
            encode_leaf(&[1, 2, 3]),
            encode_leaf(&[4, 5, 6, 7]),
            encode_leaf(&[]),
            encode_leaf(&[9; 12]),
            encode_leaf(b"final"),
        ];

        let tree = Blake3MerkleTree::from_leaves(payloads.iter()).expect("tree");
        let root = tree.root();

        for (index, leaf) in payloads.iter().enumerate() {
            let path = tree.open(index).expect("path");
            verify_path(leaf, index, tree.leaf_count(), &path, &root).expect("verify");
        }
    }

    #[test]
    fn verify_fails_on_wrong_sibling_order() {
        let payloads = vec![
            encode_leaf(&[1, 2, 3, 4]),
            encode_leaf(&[5, 6, 7, 8]),
            encode_leaf(&[9, 10, 11, 12]),
            encode_leaf(&[13, 14, 15, 16]),
        ];

        let tree = Blake3MerkleTree::from_leaves(payloads.iter()).expect("tree");
        let mut path = tree.open(2).expect("path");
        // Corrupt the sibling digest to break the order constraint.
        path[0].siblings[0][0] ^= 0x01;
        let err = verify_path(&payloads[2], 2, tree.leaf_count(), &path, &tree.root()).unwrap_err();
        assert_eq!(err, MerkleError::ErrMerkleSiblingOrder);
        assert_eq!(err.to_string(), "ErrFRIPathInvalid: sibling order");
    }

    #[test]
    fn verify_fails_on_bad_index_byte() {
        let payloads = vec![encode_leaf(&[42]), encode_leaf(&[43])];
        let tree = Blake3MerkleTree::from_leaves(payloads.iter()).expect("tree");
        let mut path = tree.open(0).expect("path");
        path[0].index = MerkleIndex(2);
        let err = verify_path(&payloads[0], 0, tree.leaf_count(), &path, &tree.root()).unwrap_err();
        assert_eq!(err, MerkleError::ErrPathIndexByte);
        assert_eq!(err.to_string(), "ErrFRIPathInvalid: index byte");
    }

    #[test]
    fn verify_fails_on_leaf_len_mismatch() {
        let mut leaf = encode_leaf(&[1, 2, 3]);
        leaf[0] ^= 0x01; // corrupt the length prefix
        let path = Vec::new();
        let err = compute_root_from_path(&leaf, 0, 1, &path).unwrap_err();
        assert_eq!(err, MerkleError::ErrMerkleLeafLength);
        assert_eq!(err.to_string(), "ErrFRIPathInvalid: leaf length");
    }

    #[test]
    fn verify_fails_on_bad_padding() {
        let leaf = encode_leaf(&[42]);
        let path = vec![MerklePathElement {
            index: MerkleIndex(0),
            siblings: [[0xFF; DIGEST_SIZE]],
        }];
        let err = compute_root_from_path(&leaf, 0, 1, &path).unwrap_err();
        assert_eq!(err, MerkleError::ErrMerkleEmptyPadding);
        assert_eq!(err.to_string(), "ErrFRIPathInvalid: right padding");
    }

    #[test]
    fn right_padding_with_empty_constant_ok() {
        assert_eq!(EMPTY_DIGEST, hash(b"RPP-MERKLE-EMPTY\0").into_bytes());
        let payloads = vec![
            encode_leaf(&[1]),
            encode_leaf(&[2]),
            encode_leaf(&[3]),
            encode_leaf(&[4]),
            encode_leaf(&[5]),
        ];
        let tree = Blake3MerkleTree::from_leaves(payloads.iter()).expect("tree");
        let index = payloads.len() - 1;
        let path = tree.open(index).expect("path");
        assert!(path.iter().any(|element| element
            .siblings
            .iter()
            .any(|sibling| sibling == &EMPTY_DIGEST)));
        verify_path(
            &payloads[index],
            index,
            tree.leaf_count(),
            &path,
            &tree.root(),
        )
        .expect("verification");
    }
}
