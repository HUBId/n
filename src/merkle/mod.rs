//! Canonical Merkle commitment layer for `rpp-stark`.
//!
//! The module fixes the following protocol knobs:
//!
//! * **Arity:** Binary (`2`) and quaternary (`4`) trees are supported.  Missing
//!   children on the rightmost edge are deterministically duplicated from the
//!   last present child (Rightmost-Child Duplication â€“ RMD).
//! * **Leaf layout:** leaves are encoded as the concatenation of `leaf_width`
//!   field elements in little-endian byte order.  The order is always
//!   row-major.  No length prefix or index tag is added.
//! * **Domain separation:** every hash invocation receives the `domain_sep`
//!   parameter from [`StarkParams`](crate::params::StarkParams), preceded by a one byte node tag (`0x00`
//!   for leaves, `0x01` for internal nodes).
//! * **Hash family:** the hashing backend is selected through
//!   [`StarkParams::hash`](crate::params::StarkParams::hash).  The [`MerkleHasher`] trait abstracts the concrete
//!   implementation to keep the commitment layer family-neutral.
//!
//! The public API re-exports the most relevant types for convenience.

mod deterministic;
mod proof;
mod ser;
pub mod traits;
mod tree;
mod types;

pub use proof::{verify_proof, MerkleProof, ProofBuilder};
pub use ser::{decode_commit_aux, decode_proof, encode_commit_aux, encode_proof};
pub use traits::{MerkleCommit, MerkleHasher};
pub use tree::{CommitAux, MerkleTree};
pub use types::{Digest, EndianEncoding, Leaf, MerkleArityExt, MerkleError, ProofNode, SerKind};

pub use deterministic::DeterministicMerkleHasher;
pub use types::{Node, TreeDepth};
