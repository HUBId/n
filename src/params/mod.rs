#![forbid(unsafe_code)]

//! Canonical parameter registry for the STARK pipeline.
//!
//! This module defines [`StarkParams`] as the single source of truth for every
//! security and performance relevant parameter in the proving stack.  The
//! structure is intentionally explicit and designed to be serialised into a
//! deterministic byte layout that is shared between the prover, verifier and
//! tooling such as auditors or snapshot tests.
//!
//! # Overview
//!
//! The parameter space is split into themed sub-structures and enums.  The
//! following table summarises the high level groupings:
//!
//! | Group | Description |
//! |-------|-------------|
//! | Field & Hash | Selection of the prime field and hash function family used throughout the pipeline. |
//! | LDE & FRI | Low Degree Extension and Fast Reed–Solomon IOP knobs governing blowup factors and folding strategies. |
//! | Merkle & Transcript | Commitment encoding rules and Fiat–Shamir transcript framing. |
//! | Proof & Security | Envelope size limits and overall soundness budgeting. |
//!
//! All validation logic lives in [`validate`], canonical serialisation lives in
//! [`ser`] and the stable digest computation is exposed via
//! [`StarkParams::params_hash`].  Consumers are expected to use the
//! [`StarkParamsBuilder`] helper which offers safe defaults and pre-defined
//! [`BuiltinProfile`] presets.
//!
//! # Invariants
//!
//! * The canonical serialisation is strictly ordered as documented in
//!   [`ser`] and is stable across Rust versions.
//! * [`StarkParams::params_hash`] commits to this serialisation and therefore
//!   uniquely identifies compatible parameter sets.
//! * [`StarkParams::is_compatible_with`] only allows variations on
//!   non-critical fields, permitting for example different proof size budgets
//!   while still sharing the same security assumptions.
//!
//! # Examples
//!
//! While the module intentionally avoids executable examples to keep the
//! specification deterministic, the tests and snapshot fixtures demonstrate how
//! `BuiltinProfile::PROFILE_X8` and `BuiltinProfile::PROFILE_HISEC_X16` expand
//! into complete parameter sets.

mod builder;
mod hash;
mod ser;
mod stark_params;
mod types;
mod validate;

pub use crate::ser::{SerError, SerKind};
pub use builder::{BuiltinProfile, StarkParamsBuilder};
pub use hash::params_hash;
pub use ser::{deserialize_params, serialize_params};
pub use stark_params::StarkParams;
pub use types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashFamily, HashKind, LdeOrder,
    LdeParams, MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
pub use validate::{ParamsError, ValidationReport};
