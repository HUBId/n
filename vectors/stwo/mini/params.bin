0100010003200008000000013130534d415241500440000c000105010102315241505f4c4b4d4f52505f4b5254535250502d535441524b2d50524f46494c452d58385f5f5f5f5f5f5f5f5f5f5f301040010000020000600010
