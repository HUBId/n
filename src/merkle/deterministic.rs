use crate::hash::deterministic::Hasher;
use crate::hash::merkle::{LEAF_DOMAIN_TAG, NODE_DOMAIN_TAG};
use crate::params::HashFamily;
use std::convert::TryInto;

use super::traits::MerkleHasher;

/// Deterministic pseudo BLAKE-based Merkle hasher used for testing.
pub struct DeterministicMerkleHasher;

impl MerkleHasher for DeterministicMerkleHasher {
    type Digest = [u8; 32];

    fn hash_leaves_with_tag(
        leaf_domain_tag: u8,
        domain_sep: u64,
        ordered_leaf_bytes: &[u8],
    ) -> Self::Digest {
        hash_with_tag(leaf_domain_tag, domain_sep, ordered_leaf_bytes)
    }

    fn hash_nodes_with_tag(
        node_domain_tag: u8,
        domain_sep: u64,
        ordered_children: &[Self::Digest],
    ) -> Self::Digest {
        let mut hasher = Hasher::new();
        hasher.update(&[node_domain_tag]);
        hasher.update(&domain_sep.to_le_bytes());
        for digest in ordered_children {
            hasher.update(digest);
        }
        hasher.finalize().into_bytes()
    }

    fn digest_size() -> usize {
        match Self::hash_family() {
            HashFamily::Blake2s => crate::hash::merkle::DIGEST_SIZE,
            _ => unreachable!("deterministic merkle hasher only supports Blake2s"),
        }
    }

    fn from_bytes(bytes: &[u8]) -> Option<Self::Digest> {
        bytes.try_into().ok()
    }

    fn hash_family() -> HashFamily {
        HashFamily::Blake2s
    }
}

fn hash_with_tag(tag: u8, domain_sep: u64, payload: &[u8]) -> [u8; 32] {
    let mut hasher = Hasher::new();
    hasher.update(&[tag]);
    hasher.update(&domain_sep.to_le_bytes());
    hasher.update(payload);
    hasher.finalize().into_bytes()
}

impl DeterministicMerkleHasher {
    /// Returns the canonical leaf domain separation tag used for Blake2s Merkle commitments.
    pub const fn leaf_domain_tag() -> u8 {
        LEAF_DOMAIN_TAG
    }

    /// Returns the canonical node domain separation tag used for Blake2s Merkle commitments.
    pub const fn node_domain_tag() -> u8 {
        NODE_DOMAIN_TAG
    }
}
