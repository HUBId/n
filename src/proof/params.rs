use crate::config::{
    ProfileConfig, MERKLE_SCHEME_ID_BLAKE3_4ARY_V1, PROFILE_HIGH_SECURITY_CONFIG,
    PROFILE_STANDARD_CONFIG,
};
use crate::params::{FriFolding, HashKind, LdeOrder, MerkleArity, StarkParams, StarkParamsBuilder};

/// Returns the canonical Stark parameter set shared by prover and verifier.
///
/// The current implementation relies on the deterministic profile used by the
/// built-in prover.  Both parties fix the hashing backend to Blake2s and commit
/// to single-field-element Merkle leaves.  The helper mirrors the prover
/// configuration so callers can rely on identical Merkle framing without
/// leaking the builder details elsewhere in the crate.
pub fn canonical_stark_params(profile: &ProfileConfig) -> StarkParams {
    let mut builder = StarkParamsBuilder::new();
    builder.hash = HashKind::Blake2s { digest_size: 32 };
    builder.merkle.leaf_width = 1;
    if profile.merkle_scheme_id == MERKLE_SCHEME_ID_BLAKE3_4ARY_V1 {
        builder.merkle.arity = MerkleArity::Quaternary;
    } else {
        builder.merkle.arity = MerkleArity::Binary;
    }
    builder.lde.blowup = profile.lde_factor as u32;
    builder.fri.queries = profile.fri_queries;
    builder.fri.domain_log2 = profile.fri_depth_range.max as u16;
    builder.fri.num_layers = profile
        .fri_depth_range
        .max
        .saturating_sub(profile.fri_depth_range.min)
        .saturating_add(1);

    if profile.id == PROFILE_HIGH_SECURITY_CONFIG.id {
        builder.lde.order = LdeOrder::ColMajor;
        builder.fri.folding = FriFolding::Coset;
    } else if profile.id == PROFILE_STANDARD_CONFIG.id {
        builder.lde.order = LdeOrder::RowMajor;
        builder.fri.folding = FriFolding::Natural;
    }

    builder
        .build()
        .expect("canonical Stark parameters must be valid")
}
