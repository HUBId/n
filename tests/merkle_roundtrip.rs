use proptest::prelude::*;
use rpp_stark::merkle::{
    decode_commit_aux, decode_proof, encode_commit_aux, encode_proof, CommitAux,
    DeterministicMerkleHasher, Leaf, MerkleCommit, MerkleError, MerkleTree,
};
use rpp_stark::params::{Endianness, HashKind, MerkleArity, StarkParams, StarkParamsBuilder};

fn build_params(arity: MerkleArity, leaf_width: u8) -> StarkParams {
    let mut builder = StarkParamsBuilder::new();
    builder.hash = HashKind::Blake2s { digest_size: 32 };
    builder.merkle.arity = arity;
    builder.merkle.leaf_width = leaf_width;
    builder.merkle.leaf_encoding = Endianness::Little;
    builder.build().expect("valid params")
}

fn make_leaves(count: usize, width: u8) -> Vec<Leaf> {
    let mut leaves = Vec::with_capacity(count);
    for i in 0..count {
        let mut bytes = Vec::new();
        for j in 0..width as usize {
            let value = (i * width as usize + j) as u64;
            bytes.extend_from_slice(&value.to_le_bytes());
        }
        leaves.push(Leaf::new(bytes));
    }
    leaves
}

fn commit_aux(params: &StarkParams, leaves: Vec<Leaf>) -> (rpp_stark::merkle::Digest, CommitAux) {
    let mut tree = MerkleTree::<DeterministicMerkleHasher>::new(params).unwrap();
    let root = tree.commit(leaves.into_iter()).expect("commit");
    let aux = tree.into_aux();
    (root, aux)
}

#[test]
fn roundtrip_binary_single_index() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(16, params.merkle().leaf_width);
    let (root, aux) = commit_aux(&params, leaves.clone());
    let proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[3]).unwrap();
    MerkleTree::<DeterministicMerkleHasher>::verify(&params, &root, &proof, &[leaves[3].clone()])
        .unwrap();
}

#[test]
fn roundtrip_quaternary_multi_index() {
    let params = build_params(MerkleArity::Quaternary, 4);
    let leaves = make_leaves(32, params.merkle().leaf_width);
    let (root, aux) = commit_aux(&params, leaves.clone());
    let proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[1, 7, 11]).unwrap();
    let selected = vec![leaves[1].clone(), leaves[7].clone(), leaves[11].clone()];
    MerkleTree::<DeterministicMerkleHasher>::verify(&params, &root, &proof, &selected).unwrap();
}

#[test]
fn determinism_snapshot() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(8, params.merkle().leaf_width);
    let (root, aux) = commit_aux(&params, leaves.clone());
    let proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[2, 5]).unwrap();
    let proof_bytes = encode_proof(&proof).unwrap();
    let aux_bytes = encode_commit_aux(&aux).unwrap();
    insta::assert_snapshot!("merkle_root_bin", hex_bytes(root.as_bytes()));
    insta::assert_snapshot!("merkle_proof_bin", hex_bytes(&proof_bytes));
    insta::assert_snapshot!("merkle_aux_bin", hex_bytes(&aux_bytes));
}

#[test]
fn proof_serialization_roundtrip() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(8, params.merkle().leaf_width);
    let (_, aux) = commit_aux(&params, leaves.clone());
    let proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[1, 6]).unwrap();
    let encoded = encode_proof(&proof).unwrap();
    let decoded = decode_proof(&encoded).unwrap();
    assert_eq!(proof.indices, decoded.indices);
    assert_eq!(proof.path.len(), decoded.path.len());
}

#[test]
fn aux_serialization_roundtrip() {
    let params = build_params(MerkleArity::Quaternary, 4);
    let leaves = make_leaves(8, params.merkle().leaf_width);
    let (_, aux) = commit_aux(&params, leaves);
    let encoded = encode_commit_aux(&aux).unwrap();
    let decoded = decode_commit_aux(&encoded).unwrap();
    assert_eq!(aux.leaf_count, decoded.leaf_count);
    assert_eq!(aux.domain_sep, decoded.domain_sep);
}

#[test]
fn invalid_index_open() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(4, params.merkle().leaf_width);
    let (_, aux) = commit_aux(&params, leaves);
    let err = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[10]).unwrap_err();
    assert!(matches!(err, MerkleError::IndexOutOfRange { .. }));
}

#[test]
fn duplicate_index_open() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(4, params.merkle().leaf_width);
    let (_, aux) = commit_aux(&params, leaves);
    let err = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[1, 1]).unwrap_err();
    assert!(matches!(err, MerkleError::DuplicateIndex { .. }));
}

#[test]
fn tampered_proof_fails() {
    let params = build_params(MerkleArity::Binary, 4);
    let leaves = make_leaves(8, params.merkle().leaf_width);
    let (root, aux) = commit_aux(&params, leaves.clone());
    let mut proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &[2]).unwrap();
    if let Some(first) = proof.path.first_mut() {
        if let Some(sibling) = first.siblings_mut().first_mut() {
            let bytes = sibling.as_bytes_mut();
            bytes[0] ^= 0x01;
        }
    }
    let err = MerkleTree::<DeterministicMerkleHasher>::verify(
        &params,
        &root,
        &proof,
        &[leaves[2].clone()],
    )
    .unwrap_err();
    assert!(matches!(err, MerkleError::VerificationFailed));
}

fn hex_bytes(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

proptest! {
    #[test]
    fn random_roundtrip(num_leaves in 2usize..32, indices in proptest::collection::btree_set(0u32..64u32, 1..4)) {
        let params = build_params(MerkleArity::Binary, 4);
        let leaves = make_leaves(num_leaves, params.merkle().leaf_width);
        let remapped: std::collections::BTreeSet<u32> = indices
            .into_iter()
            .map(|i| i % num_leaves as u32)
            .collect();
        let indices: Vec<u32> = remapped.into_iter().collect();
        let (root, aux) = commit_aux(&params, leaves.clone());
        let proof = MerkleTree::<DeterministicMerkleHasher>::open(&params, &aux, &indices).unwrap();
        for node in proof.path() {
            let mut seen_zero = false;
            for sibling in node.siblings() {
                let is_zero = sibling.as_bytes().iter().all(|b| *b == 0);
                if is_zero {
                    seen_zero = true;
                } else {
                    assert!(!seen_zero, "non-zero digest after zero padding");
                }
            }
        }
        let selected: Vec<Leaf> = indices.iter().map(|i| leaves[*i as usize].clone()).collect();
        MerkleTree::<DeterministicMerkleHasher>::verify(&params, &root, &proof, &selected).unwrap();
    }
}
