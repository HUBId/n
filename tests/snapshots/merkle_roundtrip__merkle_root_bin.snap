---
source: tests/merkle_roundtrip.rs
assertion_line: 65
expression: hex_bytes(root.as_bytes())
---
c3e088d80ca0564bd4a57a3df3208c3f3129dd52caf9107ad8c59d9f5b80b0c3
